src/server/ApiRoutes/Search.js

```diff
@@ -68,34 +69,48 @@ function fetchResults(searchKeywords = '', page, sortBy, order, field, filters,
       .catch(console.error))
     .then((response) => {
       const [results, aggregations, drbbResults] = response;
-      return Promise.all(results.itemListElement.map((resultObj) => {
+      const locationCodes = new Set();
+      const { itemListElement } = results;
+      itemListElement.forEach((resultObj) => {
+        let numItemsProcessed = 0;
         const { result } = resultObj;
-        return addLocationUrls(result).then((updatedResult) => {
-          const { holdings } = updatedResult;
-          if (!holdings) return { ...resultObj, updatedResult };
-          addCheckInItems(updatedResult);
-          holdings.slice(0, itemTableLimit).forEach(
-            holding => addHoldingDefinition(holding));
-          return { ...resultObj, updatedResult };
+        const { holdings } = resultObj.result;
+        if (holdings) {
+          addCheckInItems(result);
+          holdings.slice(0, itemTableLimit).forEach((holding) => {
+            addHoldingDefinition(holding);
+            if (numItemsProcessed < 0 && holding.location) {
+              locationCodes.add(holding.location[0].code);
+              numItemsProcessed += 1;
+            }
+          });
+        }
+        if (numItemsProcessed < itemTableLimit) {
+          result.items.slice(0, itemTableLimit - numItemsProcessed).forEach(
+            item => locationCodes.add(item.holdingLocation[0]['@id']));
+        }
+      });
```

Comment:
This is a bit confusing. What is the intended effect?

Reply:
We only need to retrieve data for the items that will actually be displayed on the search results page

Reply:
Excuse multiple comments - the other case is that there are more than 3 holdings, therefore, we can stop adding to `locationCodes` and we also do not need to look at the items